### ISSUE TITLE
Question regarding setup time

### ISSUE THREAD

**Issue by fbaierl** [VROOM-Project/vroom#897]
I stumbled accross the following problem:

- two shipments with the exact same parameters (time, location, etc.)
- one can be assigned, the other one remains unassigned

Now I am wondering as to why this is happening. If I remove the setup time it works, but as I understood the setup time this only has to be done once for both pickups and once for both deliveries. 

So how is this possible? After the setup time is finished, it should not take any time for the second shipment to be picked up and same at the delivery point.

**Input:**

```
{
    "vehicles": [
        {
            "id": 0,
            "description": "229ebd2f-7051-474a-9290-0c5ee64e6588",
            "start": [
                11.90148,
                51.46966
            ],
            "end": [
                11.90148,
                51.46966
            ],
            "capacity": [
                6,
                0
            ],
            "skills": [
                1000,
                2,
                3,
                5
            ],
            "time_window": [
                1678329180,
                1678360500
            ],
            "breaks": [
                {
                    "id": 0,
                    "time_windows": [
                        [
                            1678338900,
                            1678349880
                        ]
                    ],
                    "service": 1800,
                    "max_load": [
                        0,
                        0
                    ]
                }
            ],
            "speed_factor": 1.1
        }
    ],
    "jobs": [],
    "shipments": [
        {
            "amount": [
                1,
                0
            ],
            "skills": [
                2,
                5
            ],
            "priority": 0,
            "pickup": {
                "id": 0,
                "setup": 105,
                "service": 0,
                "location": [
                    11.802052,
                    51.541224
                ],
                "time_windows": [
                    [
                        1678333380,
                        1678333740
                    ]
                ],
                "description": "pickup|a6fc3676-c693-4ae1-88de-39d088a7b791"
            },
            "delivery": {
                "id": 1,
                "setup": 105,
                "service": 0,
                "location": [
                    11.823286,
                    51.527426
                ],
                "time_windows": [
                    [
                        1678333380,
                        1678333740
                    ]
                ],
                "description": "delivery|a6fc3676-c693-4ae1-88de-39d088a7b791"
            }
        },
        {
            "amount": [
                1,
                0
            ],
            "skills": [
                2,
                5
            ],
            "priority": 0,
            "pickup": {
                "id": 2,
                "setup": 105,
                "service": 0,
                "location": [
                    11.802052,
                    51.541224
                ],
                "time_windows": [
                    [
                        1678333380,
                        1678333740
                    ]
                ],
                "description": "pickup|25d5a001-0871-4c8d-9b25-2777df06c574"
            },
            "delivery": {
                "id": 3,
                "setup": 105,
                "service": 0,
                "location": [
                    11.823286,
                    51.527426
                ],
                "time_windows": [
                    [
                        1678333380,
                        1678333740
                    ]
                ],
                "description": "delivery|25d5a001-0871-4c8d-9b25-2777df06c574"
            }
        }
    ]
}
```

**Output**

```
{
    "code": 0,
    "summary": {
        "cost": 1952,
        "routes": 1,
        "unassigned": 2,
        "delivery": [
            1,
            0
        ],
        "amount": [
            1,
            0
        ],
        "pickup": [
            1,
            0
        ],
        "setup": 210,
        "service": 1800,
        "duration": 1952,
        "waiting_time": 4298,
        "priority": 0,
        "distance": 31906,
        "violations": [],
        "computing_times": {
            "loading": 17,
            "solving": 0,
            "routing": 5
        }
    },
    "unassigned": [
        {
            "id": 3,
            "location": [
                11.823286,
                51.527426
            ],
            "type": "delivery",
            "description": "delivery|25d5a001-0871-4c8d-9b25-2777df06c574"
        },
        {
            "id": 2,
            "location": [
                11.802052,
                51.541224
            ],
            "type": "pickup",
            "description": "pickup|25d5a001-0871-4c8d-9b25-2777df06c574"
        }
    ],
    "routes": [
        {
            "vehicle": 0,
            "cost": 1952,
            "description": "229ebd2f-7051-474a-9290-0c5ee64e6588",
            "delivery": [
                1,
                0
            ],
            "amount": [
                1,
                0
            ],
            "pickup": [
                1,
                0
            ],
            "setup": 210,
            "service": 1800,
            "duration": 1952,
            "waiting_time": 4298,
            "priority": 0,
            "distance": 31906,
            "steps": [
                {
                    "type": "start",
                    "location": [
                        11.90148,
                        51.46966
                    ],
                    "setup": 0,
                    "service": 0,
                    "waiting_time": 0,
                    "load": [
                        0,
                        0
                    ],
                    "arrival": 1678332440,
                    "duration": 0,
                    "violations": [],
                    "distance": 0
                },
                {
                    "type": "pickup",
                    "description": "pickup|a6fc3676-c693-4ae1-88de-39d088a7b791",
                    "location": [
                        11.802052,
                        51.541224
                    ],
                    "id": 0,
                    "setup": 105,
                    "service": 0,
                    "waiting_time": 0,
                    "job": 0,
                    "load": [
                        1,
                        0
                    ],
                    "arrival": 1678333406,
                    "duration": 966,
                    "violations": [],
                    "distance": 15994
                },
                {
                    "type": "delivery",
                    "description": "delivery|a6fc3676-c693-4ae1-88de-39d088a7b791",
                    "location": [
                        11.823286,
                        51.527426
                    ],
                    "id": 1,
                    "setup": 105,
                    "service": 0,
                    "waiting_time": 0,
                    "job": 1,
                    "load": [
                        0,
                        0
                    ],
                    "arrival": 1678333740,
                    "duration": 1195,
                    "violations": [],
                    "distance": 18878
                },
                {
                    "type": "break",
                    "location_index": 0,
                    "id": 0,
                    "setup": 0,
                    "service": 1800,
                    "waiting_time": 4298,
                    "load": [
                        0,
                        0
                    ],
                    "arrival": 1678334602,
                    "duration": 1952,
                    "violations": [],
                    "distance": 31906
                },
                {
                    "type": "end",
                    "location": [
                        11.90148,
                        51.46966
                    ],
                    "setup": 0,
                    "service": 0,
                    "waiting_time": 0,
                    "load": [
                        0,
                        0
                    ],
                    "arrival": 1678340700,
                    "duration": 1952,
                    "violations": [],
                    "distance": 31906
                }
            ],
            "violations": [],
            "geometry": "{vcyH_msgAYQiCqBgA}@YYCHqA~DaA|C[jAa@rAEJIZeBkB}@cAiAqAOQc@g@[_@s@w@QSe@e@K\\AHGZId@Kt@QjAa@vBId@ATAR?j@AZ?LD`@Jr@TbAP~@H`AAbAKf@GTG^E\\Gf@k@xBy@nDoAxFcDdOsCdNk@`Ck@jBe@lAg@lAqBhE_ChFs@hBCFUh@m@fBg@hBOd@GXWfA_@bBm@hDm@xDa@fD[dD]~D{@nOq@rMm@~K[vFWfFO~CEtBCrDDlDJ|CR|C`AzMX~Df@dHJzAFnBD`D@fABlDFvCT`GN`D`@`KNrDVjGTnF@\\B\\Bv@`@lJ\\nJF~@RrELnBPlB^zC^pBTdAXfA^tAj@fB`BbFp@tB~@dDf@dBh@lBbAzEnAxHd@lEZfDZvFLbEFhE?rDA|CA`A]Ia@IeCc@UEiGmA_Dm@_AK{@AK@MFKLGRI`@Kn@Ml@ABMj@Ol@]nAa@~Ai@a@_Ae@q@W}B_AeB[ICeCi@aDo@}Dy@cASGAeGiAgJkBaGgAC?wSeD_BGk@A{@@iBJcBPmBLWB]JSNEDIHQ\\ITGXGd@GnAGvAMxA[pCKx@IZELCJMXSZUV]XUTU^e@`A_@`AQl@Mj@Mx@Gv@IbAEd@m@`HK|AG`B_@zIGnDOzAIhAEbAIhBEr@K`@U\\QJKBe@B{@UaBe@aBQ_A[o@[KGKE_B}@{@c@eGeDWMyC_BUM{[_QkE_CiB_AGCaCqAgAk@OIMGGGYQg@c@y@cAc@m@uAmBaB{BYS_Ai@c@Yc@]SQUSWWm@e@k@Om@GUCaAYEAUKoAk@MG}@_@{@a@QIo@]_EmBqAk@i@SYIMC_AO_AMeC_@wA[i@Qq@Sa@S_EgCoBmAk@]yAgAe@[UKMEK?GDCNE^ShBKt@KbAE^ARa@pDKbAE^Ix@UtAMb@MZoBnCo@|@sCtDS^m@lAQZqAhC[t@]z@GRKn@K`AKnBGfACtA@dEAxD?@?t@GdCK|AS|AsPnfAS@qRjDkDd@wGjAgHrAsEx@{Ah@o@@c@Ee@m@Ii@GcAAYBsAFiAAuAGAF@@tAGhACrA@XFbAHh@d@l@b@Dn@AzAi@rEy@fHsAvGkAjDe@pRkDRArPofAR}AJ}AFeC?u@?A@yDAeEBuAFgAJoBJaAJo@FS\\{@Zu@pAiCP[l@mAR_@rCuD?{AIaDM?k@Bg@A?N@VBTCUAW?Of@@j@CL?vCBNBNJLc@TuAHy@D_@JcA`@qD@SD_@JcAJu@RiBD_@BOFEJ?LDTJd@ZxAfAj@\\nBlA~DfC`@Rp@Rh@PvAZdC^~@L~@NLBXHh@RpAj@~DlBn@\\PHz@`@|@^LFnAj@TJD@`AXTBl@Fj@Nl@d@VVTRRPb@\\b@X~@h@XR`BzBtAlBb@l@x@bAf@b@XPFFLFNHfAj@`CpAFBhB~@jE~Bz[~PTLxC~AVLdGdDz@b@~A|@JDJFn@Z~@Z`BP`Bd@z@Td@CJCPKT]Ja@Ds@HiBDcAHiAN{AFoD^{IFaBJ}Al@aHDe@HcAFw@Ly@Lk@Pm@^aAd@aAT_@TU\\YTWR[LYBKDMH[Jy@ZqCLyAFwAFoAFe@FYHUP]HIDERO\\KVClBMbBQhBKz@Aj@@~AFvSdDB?`GfAfJjBdGhAF@bAR|Dx@`Dn@dCh@HBdBZ|B~@p@V~@d@h@`@`@_B\\oANm@Lk@@CLm@Jo@Ha@FSJMLGJAz@@~@J~Cl@hGlATDdCb@`@H\\HRBDkF?sDGgEMeE[sF[mDe@gEoA}HgAyEc@eB}AqFiB{FeAaDe@_BWaAQo@[}AO}@Ms@]}CQuBGgAC[MyCCi@C]IoBUqFi@cMCo@_@aJE}@UuFOiF?GQeHSqIIwCSsFEu@OaDOyCa@iHYcE_AiMSgDG_DG_D@cEBeBJeDRaFZwFl@{Kr@_Nz@oOZuDLqALsAb@iDh@qDn@iD`@oBd@iBp@aCx@_CtAeDv@cBxA_DvA}Ct@mBr@_Cl@eCjBcJjEiSjAoFr@}Cr@sC^o@Vm@b@k@VOPGf@Ef@Aj@ENGRIRUXe@d@gAZaAh@wATm@DIBKBIJ]H[DK`@sAZkA`A}CpA_EBIXXfA|@hCpBXP"
        }
    ]
}
```

Related old question of mine: https://github.com/VROOM-Project/vroom/issues/768 but the situation is different this time because the time-windows are big enough.

**Comment by jcoupey** [comment]
Let's do the math in the most optimistic situation where the first pickup happens ASAP. Based on your output, the `P -> D` costs are 229.

## Option `P0 -> D1` (the one you get)

ASAP arrival at `P0` is 1678333380 based on the TW, so earliest start for `D1` is 1678333380 + 105 + 229 = 1678333714 which is fine since it's below the TW deadline (1678333740).

## Option `P0 -> P2 -> D1 -> D3` (the one you hope for)

ASAP arrival at `P0` is still 1678333380 based on the TW, so earliest start for `D1` is still 1678333380 + 105 + 229 = 1678333714. This is consistent with the fact that adding `P2` does not add any setup/service time so is basically "free".

Now earliest start for `D3` happens after setup time for `D1` so it is in fact 1678333380 + 105 + 229 + 105 = 1678333819, which is a bit after the TW deadline (1678333740). So this option is not valid based on your constraints.

## TL;DR

Setting zero setup/service times means adding an extra task is basically "free" in term of total span, so it's easy to think that both deliveries should be doable *at once*. But since tasks are done sequentially, the first setup time pushes the next task(s) so the TW deadline is actually more constraining for those.

The good news is that it can only happen for the second task at the same place in this setup (all tasks after the second would have the exact same start time), so you should be able to get the desired solution by making sure the TW you provide include that extra setup length.

**Comment by fbaierl** [comment]
Ah I see. That makes sense. So in my case I should always be careful to not make the time-window smaller than 2 * setup time + length of drive.

 Thank you very much for the detailed answer!
